File: ios/domes_app/lib/data/protocol/peer_contract.dart
import 'dart:typed_data';

import 'package:domes_app/data/proto/generated/peer_drill.pb.dart';

const int _maxPeerMessageSize = 64;

PeerMessage decodePeerMessage(
  Uint8List bytes, {
  required PeerRole receiverRole,
  required PeerLifecycleState state,
}) {
  if (bytes.isEmpty || bytes.length > _maxPeerMessageSize) {
    throw const FormatException('peer message is empty or oversized');
  }
  final PeerMessage message;
  try {
    message = PeerMessage.fromBuffer(bytes);
  } on Object {
    throw const FormatException('malformed peer protobuf');
  }
  if (_hasUnknownFields(message)) {
    throw const FormatException('unknown peer protobuf field');
  }
  if (!_sameBytes(message.writeToBuffer(), bytes)) {
    throw const FormatException('noncanonical peer protobuf');
  }
  validatePeerMessage(message, receiverRole: receiverRole, state: state);
  return message;
}

void validatePeerMessage(
  PeerMessage message, {
  required PeerRole receiverRole,
  required PeerLifecycleState state,
}) {
  if (receiverRole != PeerRole.PEER_ROLE_MASTER &&
      receiverRole != PeerRole.PEER_ROLE_SLAVE) {
    throw const FormatException('role-invalid peer receiver');
  }
  if (!message.hasHeader() ||
      message.header.version != ContractVersion.CONTRACT_VERSION_1 ||
      message.header.srcPodId > 0xffff ||
      message.header.dstPodId > 0xffff ||
      (message.header.senderMac.isNotEmpty &&
          message.header.senderMac.length != 6) ||
      message.whichPayload() == PeerMessage_Payload.notSet) {
    throw const FormatException('invalid peer fields');
  }

  final bool discovery =
      message.hasBeacon() || message.hasPing() || message.hasPong();
  final bool event = message.hasTouchEvent() || message.hasTimeoutEvent();
  final PeerRole expectedSender = discovery
      ? PeerRole.PEER_ROLE_UNSPECIFIED
      : event
      ? PeerRole.PEER_ROLE_SLAVE
      : PeerRole.PEER_ROLE_MASTER;
  if (message.header.senderRole != expectedSender) {
    throw const FormatException('role-invalid peer message');
  }

  final bool fieldsValid = switch (message.whichPayload()) {
    PeerMessage_Payload.beacon ||
    PeerMessage_Payload.ping ||
    PeerMessage_Payload.pong ||
    PeerMessage_Payload.stopAll => true,
    PeerMessage_Payload.joinGame =>
      message.joinGame.assignedRole == PeerRole.PEER_ROLE_SLAVE,
    PeerMessage_Payload.armTouch =>
      message.armTouch.roundToken != 0 &&
          message.armTouch.timeoutMs > 0 &&
          message.armTouch.timeoutMs <= 60000 &&
          message.armTouch.feedbackMode <= 3,
    PeerMessage_Payload.setColor =>
      message.setColor.r <= 255 &&
          message.setColor.g <= 255 &&
          message.setColor.b <= 255,
    PeerMessage_Payload.simulateTouch =>
      message.simulateTouch.roundToken != 0 &&
          message.simulateTouch.padIndex <= 3,
    PeerMessage_Payload.touchEvent =>
      message.touchEvent.roundToken != 0 && message.touchEvent.padIndex <= 3,
    PeerMessage_Payload.timeoutEvent => message.timeoutEvent.roundToken != 0,
    PeerMessage_Payload.notSet => false,
  };
  if (!fieldsValid) throw const FormatException('invalid peer fields');

  final bool allowed =
      (discovery &&
          state != PeerLifecycleState.PEER_LIFECYCLE_STATE_STOPPED &&
          state != PeerLifecycleState.PEER_LIFECYCLE_STATE_UNSPECIFIED) ||
      switch (state) {
        PeerLifecycleState.PEER_LIFECYCLE_STATE_DISCOVERY =>
          receiverRole == PeerRole.PEER_ROLE_SLAVE && message.hasJoinGame(),
        PeerLifecycleState.PEER_LIFECYCLE_STATE_READY =>
          message.hasStopAll() ||
              (receiverRole == PeerRole.PEER_ROLE_SLAVE &&
                  (message.hasArmTouch() || message.hasSetColor())) ||
              (receiverRole == PeerRole.PEER_ROLE_MASTER && event),
        PeerLifecycleState.PEER_LIFECYCLE_STATE_ARMED =>
          receiverRole == PeerRole.PEER_ROLE_SLAVE &&
              (message.hasSimulateTouch() || message.hasStopAll()),
        PeerLifecycleState.PEER_LIFECYCLE_STATE_UNSPECIFIED ||
        PeerLifecycleState.PEER_LIFECYCLE_STATE_STOPPED => false,
        _ => false,
      };
  if (!allowed) throw const FormatException('state-invalid peer message');
}

bool _hasUnknownFields(PeerMessage message) {
  if (message.unknownFields.isNotEmpty ||
      (message.hasHeader() && message.header.unknownFields.isNotEmpty)) {
    return true;
  }
  return switch (message.whichPayload()) {
    PeerMessage_Payload.beacon => message.beacon.unknownFields.isNotEmpty,
    PeerMessage_Payload.ping => message.ping.unknownFields.isNotEmpty,
    PeerMessage_Payload.pong => message.pong.unknownFields.isNotEmpty,
    PeerMessage_Payload.joinGame => message.joinGame.unknownFields.isNotEmpty,
    PeerMessage_Payload.armTouch => message.armTouch.unknownFields.isNotEmpty,
    PeerMessage_Payload.setColor => message.setColor.unknownFields.isNotEmpty,
    PeerMessage_Payload.stopAll => message.stopAll.unknownFields.isNotEmpty,
    PeerMessage_Payload.simulateTouch =>
      message.simulateTouch.unknownFields.isNotEmpty,
    PeerMessage_Payload.touchEvent =>
      message.touchEvent.unknownFields.isNotEmpty,
    PeerMessage_Payload.timeoutEvent =>
      message.timeoutEvent.unknownFields.isNotEmpty,
    PeerMessage_Payload.notSet => false,
  };
}

bool _sameBytes(List<int> left, List<int> right) {
  if (left.length != right.length) return false;
  for (var index = 0; index < left.length; index++) {
    if (left[index] != right[index]) return false;
  }
  return true;
}

File: tools/domes-cli/src/protocol/peer_contract.rs
//! Strict portable peer/drill protobuf consumer.

use crate::proto::peer::{
    peer_message::Payload, ContractVersion, PeerLifecycleState, PeerMessage, PeerRole,
};
use prost::Message;

const MAX_PEER_MESSAGE_SIZE: usize = 64;

#[derive(Debug, thiserror::Error, PartialEq, Eq)]
pub enum PeerContractError {
    #[error("peer message is empty or oversized")]
    Size,
    #[error("peer message protobuf is malformed or contains unknown/noncanonical fields")]
    Malformed,
    #[error("peer message has invalid fields")]
    InvalidFields,
    #[error("peer message sender role is invalid for its payload")]
    InvalidRole,
    #[error("peer message is invalid in the receiver lifecycle state")]
    InvalidState,
}

pub fn decode_peer_message(
    bytes: &[u8],
    receiver_role: PeerRole,
    state: PeerLifecycleState,
) -> Result<PeerMessage, PeerContractError> {
    if bytes.is_empty() || bytes.len() > MAX_PEER_MESSAGE_SIZE {
        return Err(PeerContractError::Size);
    }
    let message = PeerMessage::decode(bytes).map_err(|_| PeerContractError::Malformed)?;
    if message.encode_to_vec() != bytes {
        return Err(PeerContractError::Malformed);
    }
    validate_peer_message(&message, receiver_role, state)?;
    Ok(message)
}

pub fn validate_peer_message(
    message: &PeerMessage,
    receiver_role: PeerRole,
    state: PeerLifecycleState,
) -> Result<(), PeerContractError> {
    if !matches!(receiver_role, PeerRole::Master | PeerRole::Slave) {
        return Err(PeerContractError::InvalidRole);
    }
    let header = message
        .header
        .as_ref()
        .ok_or(PeerContractError::InvalidFields)?;
    if ContractVersion::try_from(header.version) != Ok(ContractVersion::ContractVersion1)
        || header.src_pod_id > u16::MAX as u32
        || header.dst_pod_id > u16::MAX as u32
        || (!header.sender_mac.is_empty() && header.sender_mac.len() != 6)
    {
        return Err(PeerContractError::InvalidFields);
    }
    let sender_role =
        PeerRole::try_from(header.sender_role).map_err(|_| PeerContractError::InvalidRole)?;
    let payload = message
        .payload
        .as_ref()
        .ok_or(PeerContractError::InvalidFields)?;

    let fields_valid = match payload {
        Payload::Beacon(_) | Payload::Ping(_) | Payload::Pong(_) | Payload::StopAll(_) => true,
        Payload::JoinGame(value) => value.assigned_role == PeerRole::Slave as i32,
        Payload::ArmTouch(value) => {
            value.round_token != 0
                && (1..=60_000).contains(&value.timeout_ms)
                && value.feedback_mode <= 3
        }
        Payload::SetColor(value) => value.r <= 255 && value.g <= 255 && value.b <= 255,
        Payload::SimulateTouch(value) => value.round_token != 0 && value.pad_index <= 3,
        Payload::TouchEvent(value) => value.round_token != 0 && value.pad_index <= 3,
        Payload::TimeoutEvent(value) => value.round_token != 0,
    };
    if !fields_valid {
        return Err(PeerContractError::InvalidFields);
    }

    let discovery = matches!(
        payload,
        Payload::Beacon(_) | Payload::Ping(_) | Payload::Pong(_)
    );
    let event = matches!(payload, Payload::TouchEvent(_) | Payload::TimeoutEvent(_));
    let expected_sender = if discovery {
        PeerRole::Unspecified
    } else if event {
        PeerRole::Slave
    } else {
        PeerRole::Master
    };
    if sender_role != expected_sender {
        return Err(PeerContractError::InvalidRole);
    }

    let allowed = discovery
        && matches!(
            state,
            PeerLifecycleState::Discovery | PeerLifecycleState::Ready | PeerLifecycleState::Armed
        )
        || match state {
            PeerLifecycleState::Discovery => {
                receiver_role == PeerRole::Slave && matches!(payload, Payload::JoinGame(_))
            }
            PeerLifecycleState::Ready => {
                matches!(payload, Payload::StopAll(_))
                    || (receiver_role == PeerRole::Slave
                        && matches!(payload, Payload::ArmTouch(_) | Payload::SetColor(_)))
                    || (receiver_role == PeerRole::Master && event)
            }
            PeerLifecycleState::Armed => {
                receiver_role == PeerRole::Slave
                    && matches!(payload, Payload::SimulateTouch(_) | Payload::StopAll(_))
            }
            PeerLifecycleState::Unspecified | PeerLifecycleState::Stopped => false,
        };
    if !allowed {
        return Err(PeerContractError::InvalidState);
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::proto::peer::{
        ArmTouch, Beacon, JoinGame, PeerHeader, Ping, Pong, SetColor, SimulateTouch, StopAll,
        TimeoutEvent, TouchEvent,
    };

    fn message(payload: Payload, role: PeerRole) -> PeerMessage {
        PeerMessage {
            header: Some(PeerHeader {
                version: ContractVersion::ContractVersion1 as i32,
                src_pod_id: 1,
                dst_pod_id: 2,
                sender_role: role as i32,
                timestamp_us: 1,
                sequence: 1,
                sender_mac: vec![1, 2, 3, 4, 5, 6],
            }),
            payload: Some(payload),
        }
    }

    #[test]
    fn generated_variants_round_trip_without_independent_contract() {
        let cases = [
            (
                Payload::Beacon(Beacon {}),
                PeerRole::Unspecified,
                PeerRole::Master,
                PeerLifecycleState::Discovery,
            ),
            (
                Payload::Ping(Ping {}),
                PeerRole::Unspecified,
                PeerRole::Master,
                PeerLifecycleState::Discovery,
            ),
            (
                Payload::Pong(Pong {}),
                PeerRole::Unspecified,
                PeerRole::Slave,
                PeerLifecycleState::Discovery,
            ),
            (
                Payload::JoinGame(JoinGame {
                    assigned_role: PeerRole::Slave as i32,
                }),
                PeerRole::Master,
                PeerRole::Slave,
                PeerLifecycleState::Discovery,
            ),
            (
                Payload::ArmTouch(ArmTouch {
                    round_token: 1,
                    timeout_ms: 3000,
                    feedback_mode: 3,
                }),
                PeerRole::Master,
                PeerRole::Slave,
                PeerLifecycleState::Ready,
            ),
            (
                Payload::SetColor(SetColor { r: 1, g: 2, b: 3 }),
                PeerRole::Master,
                PeerRole::Slave,
                PeerLifecycleState::Ready,
            ),
            (
                Payload::StopAll(StopAll {}),
                PeerRole::Master,
                PeerRole::Slave,
                PeerLifecycleState::Ready,
            ),
            (
                Payload::SimulateTouch(SimulateTouch {
                    round_token: 1,
                    pad_index: 0,
                }),
                PeerRole::Master,
                PeerRole::Slave,
                PeerLifecycleState::Armed,
            ),
            (
                Payload::TouchEvent(TouchEvent {
                    round_token: 1,
                    reaction_time_us: 100,
                    pad_index: 0,
                }),
                PeerRole::Slave,
                PeerRole::Master,
                PeerLifecycleState::Ready,
            ),
            (
                Payload::TimeoutEvent(TimeoutEvent { round_token: 1 }),
                PeerRole::Slave,
                PeerRole::Master,
                PeerLifecycleState::Ready,
            ),
        ];
        for (payload, sender, receiver, state) in cases {
            let expected = message(payload, sender);
            let bytes = expected.encode_to_vec();
            assert_eq!(
                decode_peer_message(&bytes, receiver, state).unwrap(),
                expected
            );
        }
    }

    #[test]
    fn malformed_unknown_truncated_and_oversized_fail_closed() {
        let valid = message(Payload::Ping(Ping {}), PeerRole::Unspecified).encode_to_vec();
        assert_eq!(
            decode_peer_message(
                &valid[..valid.len() - 1],
                PeerRole::Master,
                PeerLifecycleState::Discovery
            ),
            Err(PeerContractError::Malformed)
        );
        let mut unknown = valid.clone();
        unknown.extend_from_slice(&[0xf8, 0x01, 0x01]);
        assert_eq!(
            decode_peer_message(&unknown, PeerRole::Master, PeerLifecycleState::Discovery),
            Err(PeerContractError::Malformed)
        );
        assert_eq!(
            decode_peer_message(&[0; 65], PeerRole::Master, PeerLifecycleState::Discovery),
            Err(PeerContractError::Size)
        );
    }

    #[test]
    fn role_and_state_invalid_messages_fail_closed() {
        let invalid_role = message(
            Payload::TimeoutEvent(TimeoutEvent { round_token: 1 }),
            PeerRole::Master,
        )
        .encode_to_vec();
        assert_eq!(
            decode_peer_message(&invalid_role, PeerRole::Master, PeerLifecycleState::Ready),
            Err(PeerContractError::InvalidRole)
        );
        let invalid_state = message(
            Payload::ArmTouch(ArmTouch {
                round_token: 1,
                timeout_ms: 3000,
                feedback_mode: 3,
            }),
            PeerRole::Master,
        )
        .encode_to_vec();
        assert_eq!(
            decode_peer_message(&invalid_state, PeerRole::Slave, PeerLifecycleState::Armed),
            Err(PeerContractError::InvalidState)
        );
        let discovery = message(Payload::Ping(Ping {}), PeerRole::Unspecified).encode_to_vec();
        assert_eq!(
            decode_peer_message(
                &discovery,
                PeerRole::Unspecified,
                PeerLifecycleState::Discovery
            ),
            Err(PeerContractError::InvalidRole)
        );
        assert_eq!(
            decode_peer_message(
                &discovery,
                PeerRole::Master,
                PeerLifecycleState::Unspecified
            ),
            Err(PeerContractError::InvalidState)
        );
    }
}
